vault/ssorolecredentialsprovider.go

```diff
@@ -201,3 +211,191 @@ func (p *SSORoleCredentialsProvider) newOIDCToken(ctx context.Context) (*ssooidc
 		return t, nil
 	}
 }
+
+// newOIDCTokenPKCE generates a new OIDC token using the "Authorization Code Grant" flow with PKCE.
+func (p *SSORoleCredentialsProvider) newOIDCTokenPKCE(ctx context.Context) (*ssooidc.CreateTokenOutput, error) {
+	// ref: https://datatracker.ietf.org/doc/html/rfc7636
+
+	// generate a random 32 byte code verifier; base64 encode it
+	codeVerifierBytes := make([]byte, 32)
+	n, err := crand.Read(codeVerifierBytes)
+	if err != nil || n != 32 {
+		return nil, fmt.Errorf("failed to generate PKCE verifier: %w", err)
+	}
+	codeVerifier := base64.RawURLEncoding.EncodeToString(codeVerifierBytes)
+
+	// generate the code challenge: base64(sha256(codeVerifier))
+	codeChallengeBytes := sha256.Sum256([]byte(codeVerifier))
+	codeChallenge := base64.RawURLEncoding.EncodeToString(codeChallengeBytes[:])
+	log.Printf("Generated PKCE code_challenge: %q", codeChallenge)
```

Comment:
I don't think this necessarily needs to be displayed by default

Reply:
I was confused at first also at the use of `log.Printf` in this project but these log lines are not displayed unless the user passes `--debug` as they're otherwise sent to `io.Discard`
https://github.com/99designs/aws-vault/blob/d4706c87383ca9ac8492c48e2d8a84dbd7e6cc07/cli/global.go#L166-L168

Reply:
ah, good point